### DISCUSSION TITLE
ui.select breaks on certain values

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2074] [Q&A]
### Question

I found something very strange with ui.select.  It breaks on certain values, namely 10-digit timestamps like 1698147721.

# Behavior
1. Select initially appears ok.  It appears with label 'age' and default value 'age'.  
<img width="264" alt="Screen Shot 2023-11-23 at 11 43 48 AM" src="https://github.com/zauberzeug/nicegui/assets/34918456/d821f857-bda0-46ae-8949-e0ac6aa9f938">

2. Dropdown opens correctly, showing the various options.  You can select any in the list.  Value is a 10-digit timstamp, eg 30 days = 1698147721 = current timestamp minus 30 days.
<img width="309" alt="Screen Shot 2023-11-23 at 11 44 13 AM" src="https://github.com/zauberzeug/nicegui/assets/34918456/f48c1302-59c0-43fb-a3fb-a4f837c8c603">

3. Selecting any option other than default causes control to break.  Control shows empty value and no label.  Bound value `filters ['age']` reports as `none` at this time.  
<img width="190" alt="Screen Shot 2023-11-23 at 11 44 31 AM" src="https://github.com/zauberzeug/nicegui/assets/34918456/37c5d1f2-cdd1-4b79-8ae2-7641a710f947">

4. Opening dropdown again shows all the options correctly.  But selecting any of them triggers same issue.  
<img width="304" alt="Screen Shot 2023-11-23 at 11 46 25 AM" src="https://github.com/zauberzeug/nicegui/assets/34918456/92845dc2-aa86-4657-889b-1f557be9a2c1">

# Code
Here's a minimal reproducible example:
```py
import time
from nicegui import ui
from universals import *

# ------------------------------------------------
def get_timeopts (label) :

    now = int (time.time ())
    opts = {
        0           :  label ,
        now - MONTH  :  '30 days' ,
        now - WEEK   :  'week' ,
        now - DAY   :  'day' ,
        now - YEAR   :  'year' ,
    }
    return opts

# ------------------------------------------------
filters = {}

def add_filters () :

    with ui.card ().classes ('full-width') :

        opts = get_timeopts ('age')
        print (f'{ opts = }')

        node = ui.select (opts, label = 'age', value = list (opts) [0])
        node.bind_value (filters, 'age')

# ------------------------------------------------
@ui.page ('/')
def mainpage () :

    add_filters ()

    title = 'Test Select'
    ui.run (title = title)

# ------------------------------------------------
if __name__ in { '__main__', '__mp_main__' } :
    mainpage ()

```

# Attempted Fixes
I tried the follow to try and fix the issue.  Nothing makes a difference:
1. Moved opts { } dict declaration inside add_filters function, in case there was some weird scoping / garbage collection issue going on.  No difference.
2. Changed opts dict keys from int to str, in case ui.select somehow couldn't handle timestamp ints (10-digit ints like 1698147721, fits in 32 bit unsigned).  No difference.
4. Tried with and without passing value = parameter to ui.select.  No difference.

Same code works perfectly fine if I replace the keys of opts with 1,2,3,4 instead of timestamps.  So the problem seems particular to 10-digit values.  I could use 1,2,3,4 etc and convert to timestamps when accessing filters ['age'], but that seems error prone and a pain.

I don't see any reason ui.select shouldn't handle 10-digit values.  My other app (html - css - js) handles the same timestamp values in `<select> / <option>` tags just fine.  So it's not a problem with the browser environment.  The issue is somewhere in nicegui stack (quasar if I had to guess).

Anyone have any ideas what the problem is?
Thanks for your help.

**Comment by falkoschindler** [comment]
Hi @ed2050, unfortunately your code is neither minimal or reproducible. You're importing `*` from `universal`. Since it affects the options dictionary, it is pretty relevant for reproducing the issue. Or maybe it isn't and we can simply use any 10-digit number. Then it would be great to reduce the code to the bare minimum, possible a single `ui.select` element with two 10-digit options. Thanks!

**Reply by ed2050** [reply]
Sorry I missed one.  Hard to catch everything from your environment.

Revised code which still causes the error:
```py
import time
from nicegui import ui
#from universals import *

SEC   =  1
MIN   =  60 * SEC
HOUR  =  60 * MIN
DAY   =  24 * HOUR
WEEK  =   7 * DAY
MONTH =  30 * DAY
YEAR  = 365 * DAY

# ------------------------------------------------
def get_timeopts (label) :

    now = int (time.time ())
    opts = {
        0           :  label ,
        now - MONTH  :  '30 days' ,
        now - WEEK   :  'week' ,
        now - DAY   :  'day' ,
        now - YEAR   :  'year' ,
    }
    return opts

# ------------------------------------------------
filters = {}

def add_filters () :

    with ui.card ().classes ('full-width') :

        opts = get_timeopts ('age')
        print (f'{ opts = }')

        node = ui.select (opts, label = 'age', value = list (opts) [0])
        node.bind_value (filters, 'age')

# ------------------------------------------------
@ui.page ('/')
def mainpage () :

    add_filters ()

    title = 'Test Select'
    ui.run (title = title)

# ------------------------------------------------
if __name__ in { '__main__', '__mp_main__' } :
    mainpage ()
```

**Reply by falkoschindler** [reply]
Are you sure you're calling `ui.run` inside `mainpage`?

**Reply by ed2050** [reply]
Yes that's right.  Honestly I didn't notice ui.run was still inside when I added the page decorator.

You can move title = and ui.run outside of mainpage () to module level.  Behavior is the same.  Still broken.

```py
@ui.page ('/')
def mainpage () :
    add_filters ()

title = 'Test Select'
ui.run (title = title)
```

Interestingly the select _does_ work if you remove the page decorator.  Don't know why decorator breaks it.  A reference / garbage collection / race issue perhaps?  Anyway removing is not a solution.  Every user needs their own filtering page with this control.

So this version works, but isn't a solution:
```py
#@ui.page ('/')
def mainpage () :
    add_filters ()

title = 'Test Select'
ui.run (title = title)
```

**Comment by ed2050** [comment]
For completeness, here are time defs:
```py
SEC   =  1
MIN   =  60 * SEC
HOUR  =  60 * MIN
DAY   =  24 * HOUR
WEEK  =   7 * DAY
MONTH =  30 * DAY
YEAR  = 365 * DAY
```
I know, month and year are approximations.  Doesn't matter for my app.

Print output showing value of opts passed to ui.select:
```
opts = {0: 'age', 1698147812: '30 days', 1700135012: 'week', 1700653412: 'day', 1669203812: 'year'}
```

**Comment by falkoschindler** [comment]
I'm still struggling to run your code. But theoretically it should boil down to this:

```py
ui.select({0: 'age', 1698147812: '30 days', 1700135012: 'week', 1700653412: 'day', 1669203812: 'year'}, label='age', value=0)
```

This is working without any problems. What are you doing differently?

**Reply by ed2050** [reply]
I just found that the page decorator seems to trigger the problem ([see post above](https://github.com/zauberzeug/nicegui/discussions/2074#discussioncomment-7653439)).  What part can't you run?  

FYI I'm still on nicegui 1.3.16.  Perhaps a change in 1.4 makes your end produce different results?

**Comment by ed2050** [comment]
I noticed something odd in my terminal.  I simplified by removing the get_timeopts () call like this:
```py
def add_filters () :

    now = int (time.time ())
    # version 1
    opts = {0: 'age', 1698164028: '30 days', 1700151228: 'week', 1700669628: 'day', 1669220028: 'year'}
    # version2
    opts = {
        0            :  'age 2' ,
        now - MONTH  :  '30 days' ,
        now - WEEK   :  'week' ,
        now - DAY    :  'day' ,
        now - YEAR   :  'year' ,
    }

    with ui.card ().classes ('full-width') :

        print (f'age { opts = }')

        node = ui.select (opts, label = 'age', value = list (opts) [0])
        node.bind_value (filters, 'age')

```

# Observations

With the first version of opts, select always works.  No issues.

With second version of opts, I see the following in terminal.  Note that opts is printed twice with slightly different values (30 days differs by 1 second, 568 vs 569).

```
cmd> python3 test/broken.py 
age  opts = {0: 'age 2', 1698165568: '30 days', 1700152768: 'week', 1700671168: 'day', 1669221568: 'year'}
NiceGUI ready to go on http://localhost:8080, and http://10.9.8.7:8080 (install netifaces to show all IPs and speedup this message)
age  opts = {0: 'age 2', 1698165569: '30 days', 1700152769: 'week', 1700671169: 'day', 1669221569: 'year'}
```
It seems that nicegui / quasar are calling add_filters twice.  I don't know why.  I do a clean start with no browser pages open.  But clearly it prints age opts = to the terminal twice.  This is before I touch anything on the opened webpage.  I've taken no actions at this point.

Sometimes I notice there's no gap.  Age opts prints twice with exactly the same values, probably because add_filters was invoked twice within 1 second.  Like in this example.  The select worked perfectly in this case.
```
cmd> python3 test/broken.py 
age  opts = {0: 'age 2', 1698166110: '30 days', 1700153310: 'week', 1700671710: 'day', 1669222110: 'year'}
NiceGUI ready to go on http://localhost:8080, and http://10.9.8.7:8080 (install netifaces to show all IPs and speedup this message)
age  opts = {0: 'age 2', 1698166110: '30 days', 1700153310: 'week', 1700671710: 'day', 1669222110: 'year'}
```

# Conclusions

The cause seems to be dynamic `ui.select` options causing a mismatch.  The page initially loads with one set of options.  Then the page generating code runs a second time for the same client, without me doing anything and not having any existing pages open (clean browser).  This causes a mismatch between what the frontend and backend see as available options (30 days : 1700152768 vs 30 days : 1700152769).  And that's when things break.

What I don't know is whether this is supposed to happen.  Is nicegui or quasar creating two copies of the page for some reason?  Is my browser sending commands that causes the page to load twice?  Could it be browser caching, an adblock extension, or something else requesting the page twice?

Frankly I have no idea.  Perhaps this gives you enough info to track it down.  Or perhaps it's destined to be browser black magic, forever lost in the ether.  I don't know.  But for now, my one takeway is this:

# Takeaway
DO NOT use dynamic select options with user pages.  Bad things happen. 😄

**Comment by falkoschindler** [comment]
I'm not sure if this is exactly the problem from your original post, but there's a fundamental flaw in your code. Here is a _minimal_ example:
```py
model = {'value': 0}

@ui.page('/')
def main_page():
    ui.select({0: 'A', int(time.time()): 'B'}).bind_value(model)
```

If you load the main page, the options are generated dynamically, i.e. include keys 0 and the current time. Whenever you change the selection to "B", the value (the current time) is written to the `model`. When you re-visit the page, the options are re-generated with the current time (which is not the same as before). But if the old time is read from `model`, there is no matching option, thus the `value` is `None`.

You didn't mention reloading the page. But maybe you have multiple browser tabs connecting to the page and interfering with each other.

**Reply by ed2050** [reply]
Thanks falko.  Example before was as minimal as I could make at the time.  The behavior disappears with some changes (such as replacing get_timeopts with a static dict or adding / removing ui.page).  It's not easy to find what's necessary and what isn't.

I didn't mention reloading the page because I wasn't reloading the page.  Unless the act of choosing an item in the dropdown causes a reload.  That's all I did.  No multiple tabs open either.

# Identifying the Problem

Your example certainly illustrates the issue.  If this is where we leave it, that's fine.  However something more is going wrong in my example above.  As in [my previous post](https://github.com/zauberzeug/nicegui/discussions/2074#discussioncomment-7653846) - the error happens even before I interact with the page.  Page is loaded twice (as evidenced by opts printing to terminal) before I touch my browser.

Underlying cause is same one you identified - changing opts values.  But it's not clear what's causing opts to change.  Nothing the user is doing.  It's triggered by either quasar / nicegui loading the page twice, or browser requesting the page twice without user intervention.

# Going forward

In any case, I know how to program around the problem now.  That's good enough for me.  However I feel this should be mentioned prominently somewhere on ui.select page (and other effected components).  One expects that "reloading the page"either 1) destroys all previous state and creates a fresh copy on front and backend, or 2) saves all the state on the page.  This behavior is in between: some state is saved (previous option values) while other state is replace (new option values).  That's a very surprising result that I feel should be documented, or it will bite more users in the future.

Anyway thannks for your help.  I'm content to close the issue now.

**Reply by falkoschindler** [reply]
Yes, there might well be another issue. The options shouldn't be generated until the page is actually visited. But the code you posted called the page function programmatically, which contained `ui.run`. I couldn't even start the app. So I'm not sure which version showed which exact behavior.

> Example before was as minimal as I could make at the time.

There's no way around a good [MRE](https://en.wikipedia.org/wiki/Minimal_reproducible_example). Minimizing the code and making sure the app works (or doesn't work) exactly as posted is the first step for problem solving. And it's usually more efficient if you do that step yourself.

That being said, let's close this issue. But feel free to create a new one if you still encounter an issue with `ui.select` or strange reloads. 🙂

**Comment by ed2050** [comment]
> Yes, there might well be another issue. The options shouldn't be generated until the page is actually visited. 

Good to know.  Behavior I saw today indicates this model is not being followed.  Maybe it's my environment, maybe it's a bug.  Will try to pin it down.

> But the code you posted called the page function programmatically, which contained `ui.run`. I couldn't even start the app. So I'm not sure which version showed which exact behavior.

All documented errors occur in last code version posted, with ui.run at module level outside mainpage().  FWIW both versions run on my machine; calling ui.run inside decorated mainpage() runs fine (even though it was unintentional and bad practice).  Maybe a difference between 1.3.16 and 1.4.x.

> Minimizing the code and making sure the app works (or doesn't work) exactly as posted is the first step for problem solving. And it's usually more efficient if you do that step yourself.

Completely agree.  Resources are not unlimited.  That includes both your time and my brainpower to understand what's going on.  We do our best. 🙂

> That being said, let's close this issue. But feel free to create a new one if you still encounter an issue with `ui.select` or strange reloads. 🙂

I think there's an issue here worth exploring, but won't post again without more definitive info.
Have a good evening.
